### PR TITLE
feat: Show object keys alphabetically rather than in jsonb style [DHIS2-7523]

### PR DESCRIPTION
**Pull request by myanghua**
Add dependency for jsonabc
Use jsonabc to sort json object

### CODE REVIEW

**Comment by benguaraldi** [comment]
Currently, the data store shows the object keys in the jsonb style, which is ordered by length of the key, so `Laos` appears before `Angola`.  We think a more human-readable style would be `Angola` before `Laos`.

**Comment by amcgee** [comment]
Hi @myanghua @benguaraldi thanks for the PR!

The `jsonabc` dependency seems superfluous here, and this isn't performant since it will parse, sort, serialize, and then parse again the JSON string.  It's better to just sort the keys after the object has been parsed, can you change it to something like this?

```js
// on line 73
const jsonLength = result.value.length
const val = sortObjectKeys(JSON.parse(result.value))

// with this as a helper functions
const sortObjectKeys = obj => (
  Object.keys(obj)
    .sort()
    .reduce((res, key) => {
      res[key] = obj[key]
      return res
    }, {})
)
```

Thanks again for the contribution!

**Comment by benguaraldi** [comment]
Thanks, @amcgee!  Good idea!

Your code only goes one level deep, but this code recurses:
```
var sortObjectKeys = obj => (
  Object.keys(obj)
    .sort()
    .reduce((res, key) => {
      if (typeof obj[key] === 'object' && obj[key] !== null)
        res[key] = sortObjectKeys(obj[key])
      else
        res[key] = obj[key]
      return res
    }, {})
)
```

@myanghua, I think this works, but could you check?

**Review by amcgee — CHANGES_REQUESTED** [review]
Nice catch @benguaraldi, thanks

**Comment by amcgee** [comment]
We should check the type of JSON before sorting, and also skip sorting for arrays, so move the `typeof` check to the top of the recursion:

```js
// on line 73
const jsonLength = result.value.length
const val = sortObjectKeys(JSON.parse(result.value))

// with this as a helper functions
const sortObjectKeys = obj => {
  if (!obj || typeof obj !== 'object' || Array.isArray(obj)) {
    return obj
  }
  return Object.keys(obj)
    .sort()
    .reduce((res, key) => {
      res[key] = sortObjectKeys(obj[key])
      return res
    }, {})
}
```

(note: I'm writing code on GitHub so this is untested ;-)

**Comment by benguaraldi** [comment]
Good change! But I think we do want to sort arrays as it might be an array of objects.

I'll revise later, unless one of you gets to it first.

Thanks!

**Comment by benguaraldi** [comment]
Okay, I think this will descend into both arrays and objects and sort any objects it finds.

```js
// on line 73
const jsonLength = result.value.length
const val = sortObjectKeys(JSON.parse(result.value))

// with this as a helper functions
const sortObjectKeys = obj => {
  if (!obj || typeof obj !== 'object') {
    return obj
  }
  return Object.keys(obj)
    .sort()
    .reduce((res, key) => {
      res[key] = sortObjectKeys(obj[key])
      return res
    }, {})
}
```

I suppose if we were willing to have a few more lines of code, we can avoid sorting the arrays while also descending into them:
```js
var sortObjectKeys = obj => {
  if (!obj || typeof obj !== 'object') {
    return obj
  }
  if (Array.isArray(obj))
    return Object.keys(obj)
      .reduce((res, key) => {
        res[key] = sortObjectKeys(obj[key])
        return res
      }, {})
  return Object.keys(obj)
    .sort()
    .reduce((res, key) => {
      res[key] = sortObjectKeys(obj[key])
      return res
    }, {})
}
```

...but given how few times that sort will actually operate, the first version makes more sense to me.  @amcgee, do you have a preference?

**Comment by amcgee** [comment]
Good catch, agreed we should sort the array members if they're objects as well.  I think this is what we need, though, because `{ 0: 'a', 1: 'b' }` (if you sort only the owned keys of an array which are already sorted) is not the same as `['a', 'b']` (which has the `Array` prototype)

```js
var sortObjectKeys = obj => {
  if (!obj || typeof obj !== 'object') {
    return obj
  }
  if (Array.isArray(obj)) {
    return obj.map(sortObjectKeys)
  }
  return Object.keys(obj)
    .sort()
    .reduce((res, key) => {
      res[key] = sortObjectKeys(obj[key])
      return res
    }, {})
}
```

**Comment by benguaraldi** [comment]
@amcgee I tested your code outside of the app, and it looks great!  Thanks!

@myanghua Could you change your PR to use [this code](https://github.com/dhis2/datastore-app/pull/16#issuecomment-760072352) as well as these lines:
```
// on line 73
const jsonLength = result.value.length
const val = sortObjectKeys(JSON.parse(result.value))
```

Note that I inadvertently switched it from `const` to `var`—I'm not sure which is more appropriate, so please change to `const` if it should be changed.

**Comment by myanghua** [comment]
@amcgee Ben and I just did another PR.  Please let us know if it looks good.  Thanks!

**Comment by benguaraldi** [comment]
@amcgee The code change is in this PR, in case that wasn't clear:
https://github.com/dhis2/datastore-app/pull/16/commits/03182820794cef81e3abd72fceba77fa9a2917ff

File: yarn.lock [hunk 1]
```diff
@@ -3276,6 +3276,11 @@ json5@^0.5.0:
   version "0.5.0"
   resolved "https://registry.yarnpkg.com/json5/-/json5-0.5.0.tgz#9b20715b026cbe3778fd769edccd822d8332a5b2"
 
+jsonabc@^2.3.1:
+  version "2.3.1"
+  resolved "https://registry.yarnpkg.com/jsonabc/-/jsonabc-2.3.1.tgz#800a1bd158fb30ace2b4613f8725adfcb040bfb4"
+  integrity sha1-gAob0Vj7MKzitGE/hyWt/LBAv7Q=
+
```
Comment (amcgee) [inline yarn.lock:3283]: You should run `yarn install` again, jsonabc isn't needed anymore

**Review by amcgee — CHANGES_REQUESTED** [review]

File: webapp/js/utils/api.js [hunk 2]
```diff
@@ -242,3 +242,20 @@ class Api
 
 export default (() =>
     new Api(API_URL).initialize())();
+
+
+const sortObjectKeys = obj => {
+    console.log(obj);
```
[the remark this answers is not in the document]
Comment (benguaraldi) [reply webapp/js/utils/api.js:248]: Oops!  That one was my fault

File: yarn.lock (hunk 1 above)
Comment (myanghua) [reply yarn.lock]: @amcgee Thanks for catching that!  I've made another [PR ](https://github.com/myanghua/datastore-app/commit/3447b3dfdcca50d3aad853d6b8a81f903fd0c146)to address these two changes.

**Comment by amcgee** [comment]
Hi @myanghua and @benguaraldi - apologies for the rigor here, we do intend to get this into DHIS 2.36 in a couple weeks.

I've opened a PR against your fork @myanghua which performs the sorting at display time, rather than at fetch time, in order to support sorted objects which have been edited during the current session.  Please have a look, and if it's ok you can merge it there I will merge this here!

**Comment by myanghua** [comment]
@amcgee Thank you, Austin.  I've done the merge.

In case you don't already know, when some node is edited, it doesn't get re-sorted until refresh even after you change it to perform sorting at display time.  It's not a problem for me though.  Thanks again.

@benguaraldi FYI.

**Comment by amcgee** [comment]
Thanks @myanghua, yes I mentioned this in the PR:

> Note that while still in the editor the keys will not be re-sorted, but once navigating away and back to the key in question it will sort again.

It would be a bigger change to re-order while editing, but this way the sorting will be applied if you switch to a different namespace and then back.

**Comment by benguaraldi** [comment]
> It would be a bigger change to re-order while editing, but this way the sorting will be applied if you switch to a different namespace and then back.

Nice!  Originally I was assuming you'd have to reload the app.

This all seems good to go to me.  Thank you both, @myanghua and @amcgee!

**Review by amcgee — APPROVED** [review]

**Comment by amcgee** [comment]
No thank you!

**Pull request merged**